src/Processors/Transforms/getSourceFromFromASTInsertQuery.cpp

```diff
@@ -35,6 +35,9 @@ Pipe getSourceFromFromASTInsertQuery(
     if (!ast_insert_query)
         throw Exception("Logical error: query requires data to insert, but it is not INSERT query", ErrorCodes::LOGICAL_ERROR);
 
+    if (ast_insert_query->infile)
+        throw Exception("Logical error: query has infile and was send directly to server", ErrorCodes::LOGICAL_ERROR);
```

Comment:
It's not a logical error. `LOGICAL_ERROR` is a special error code for exceptions that may happen only if some invariant is broken due to bug in code. It works as ordinary exception in release builds and as assertion in debug builds. 

Btw, could you please add a test that sends query with `FROM INFILE` to http interface? Test with clickhouse-local will be also useful. It's ok to add them into `02009_from_infile.sh`, search for `${CLICKHOUSE_CURL}` and `${CLICKHOUSE_LOCAL}` in functional tests if you need examples.